_posts/2022-10-27-kani-bolero.md

```diff
@@ -0,0 +1,317 @@
+
+```rust
+        let rev_index = arr.len() - index - 1;
+```
+
+and re-run the fuzzer:
+
+```
+$ cargo bolero test check_rev
+# snip ...
+#3773   REDUCE cov: 135 ft: 138 corp: 18/282b lim: 52 exec/s: 0 rss: 43Mb L: 8/32 MS: 3 ChangeByte-ChangeBinInt-EraseBytes-
+#262144 pulse  cov: 135 ft: 138 corp: 18/282b lim: 2620 exec/s: 131072 rss: 68Mb
+#524288 pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 131072 rss: 93Mb
+#1048576        pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 116508 rss: 142Mb
+```
+
+The fuzzer will run for a while without finding any failures. We can now run Kani to *prove* the assertion in our harness:
+
+```
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 0 of 338 failed (1 unreachable)
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.90419155s
+```
+
+and voila! Kani managed to prove that there none of the checks can fail for any values of the inputs.
+
+## If we have proof capability, do we still need fuzzing?
+
+One question that might arise is: is there any point in using fuzzing alongside with Kani?
+In other words, if we can prove a harness for all inputs with Kani, would we still want to use fuzzing?
+The answer is yes, due to a number of reasons. 
+
+First of all, the amount of code that Kani can analyze is typically much smaller than the code that fuzzers can handle.
+The reason is that, as explained in the [Comparison with other tools](https://model-checking.github.io/kani/tool-comparison.html#comparison-with-other-tools) section of the Kani documentation, Kani uses a technique called [*model checking*](https://en.wikipedia.org/wiki/Model_checking) in which the program is encoded as a mathematical formula that is solved using a special solver, called a [SAT solver](https://en.wikipedia.org/wiki/SAT_solver).
+The complexity of solving this formula grows exponentially with the size of the program.
+Thus, in practice, one needs to limit the size of the program for Kani, for example through reducing the size of inputs (e.g. vector length), or "[stubbing](https://en.wikipedia.org/wiki/Method_stub)" out some of the code.
+Due to this limitation, Kani is best suited for proving small pieces of code, while fuzzing, is best suited for end-to-end/integration testing.
+So in that sense, Kani and fuzzing are complementary.
+
+Second, Kani currently supports a few platforms (e.g. `x86_64` and `arm64-apple`), so testing software targeting other platforms can only be done using Bolero's fuzzing engines.
+
+Third, software has bugs (including verification tools!).
+More specifically, Kani uses multiple tools and components that include the Rust compiler's front-end, the [MIR](https://rustc-dev-guide.rust-lang.org/mir/index.html)to [GOTO](https://diffblue.github.io/cbmc//group__goto-programs.html) translation, and the solver (CBMC + Minisat), any of which may have bugs!
+So using multiple verification approaches helps provide more assurance.
```

Comment:
```suggestion
The answer is yes, because fuzzing and Kani provide complementary benefits.

Kani verifies a Rust program by symbolically analyzing its code.  This allows Kani to make mathematical statements about the expected semantics of the Rust code being verified. Kani allows you to prove that, **for all possible inputs**, the code under verification follows its specification, **assuming everything else functions correctly**.  

Fuzzing concretely executes the program under test. This gives you end-to-end confidence that **for the set of inputs generated by the fuzzer**, the code under verification follows its specification, **under real-world conditions**.

Fuzzing and Kani fit together to give more assurance than either provides on its own.
```